pyxrf/model/scan_metadata.py

```diff
@@ -0,0 +1,213 @@
+import re
+
+
+class ScanMetadataBase:
+    """
+    This is a base class. Instances of classes should be created from
+    a child class, which has meaningful implementation of the ``_gen_default_descriptions``
+    and ``_gen_print_order`` functions.
+    """
+
+    def __init__(self):
+
+        # This dictionary contains key-value pairs that represent parameter values
+        self._values = {}
+
+        # The following dictionary contains key-description pairs, that represent
+        #   user-friendly description of each key, that may be used for printing.
+        #   It is not assumed that there is one-to-one match for the keys
+        #   in 'values' and 'key_description' dictionaries. The 'key_descriptions'
+        #   should contain comprehensive list of keys. It is also assumed that
+        #   some keys in 'values' dictionary may not have matching descriptions.
+        #   'key_descriptions' dictionary set at the different place in the program.
+        #   In principle the descriptions may not be set at all.
+        self.descriptions = self._gen_default_descriptions()
+
+        self._print_order = self._gen_default_print_order()
+
+    def _gen_default_print_order(self):
+        """
+        This function must have meaningful implementation in the child class
+        """
+        return []
+
+    def _gen_default_descriptions(self):
+        """
+        This function must have meaningful implementation in the child class
+        """
+        return {}
+
+    # The following methods override standard operators to enable direct access
+    #   to ``self.values`` dictionary
+
+    def __getitem__(self, key):
+        return self._values[key]
+
+    def __setitem__(self, key, value):
+        self._values[key] = value
+
+    def __delitem__(self, key):
+        del self._values[key]
+
+    def __contains__(self, key):
+        return key in self._values
+
+    def __iter__(self):
+        return self._values.__iter__()
+
+    def keys(self):
+        return self._values.keys()
+
+    def values(self):
+        return self._values.values()
+
+    def items(self):
+        return self._values.items()
+
+    def get_metadata_dictionary(self):
+        """
+        Returns reference to metadata dictionary ``self._values``
+        """
+        return self._values
+
+    def is_metadata_available(self):
+        """
+        Returns True if at least one metadata key:value pair is available.
+        Otherwise returns False.
+        """
+        return bool(self._values)
+
+    def get_formatted_output(self):
+        """
+        Returns formatted metadata in the form ready for printing.
+        Formatting is performed based on specifications located in ``self._print_order``
+        list. Key names are replaced by descriptions from ``self.descriptions`` if
+        the available.
+        """
```

Comment:
I don't think those options will work, since the goal is not to print a dictionary as dictionary, but to generate printable representation of parameters. I think once the parameters are displayed as a table (in ``qt5`) instead of a text in a message box, this function is no longer be needed.

Reply:
👌

Reply:
Please see my comments in https://github.com/NSLS-II/PyXRF/pull/194#discussion_r341621046, which may help.

Reply:
I thought you mean replacing whole string generation function with a standard one. I used ``textwrap``.

Reply:
Yes, thanks.